### DISCUSSION TITLE
Need Pin out of cable between PET computer and M5 StickC Plus.

### DISCUSSION THREAD

**Discussion by jimdinunzio** [PlummersSoftwareLLC/PETRock#29] [Q&A] [answered]
Hi All,
I am working making a PetRock cable. 
Is this the correct pin connections on both sides, M5StickC Plus Grove conn. to PET User Port?

```
PET User Port
+-----+-------------+-----+--------+
| PIN | SIGNAL      | PIN | SIGNAL |
+-----+-------------+-----+--------+
|  1  | GND         |  A  | GND    | --> N, GND from M5
|  2  | VIDEO       |  B  | CA1(Rx)| --> Tx (pin 3 of m5stick Grove conn., ESP pin 32)
|  3  | /SRQ IN     |  C  | PA0    | --> CA1
|  4  | /EOI        |  D  |  |1    |
|  5  | DIAG        |  E  |  |2    |
|  6  | #2CASSREAD  |  F  |  |3    |
|  7  | CASSWRITE   |  H  |  |4    |
|  8  | #1CASSREAD  |  J  |  |5    |
|  9  | VERT DRIVE  |  K  |  |6    |
| 10  | HORZ DRIVE  |  L  |  V7    |
| 11  | GRAPHIC     |  M  | CB2(Tx)| --> Rx (pin 4 of m5stick Grove conn., ESP pin 33)
| 12  | GND         |  N  | GND    | --> A (GND)
+----------------------------------+
```

**Comment by davepl** [comment]
Not sure this helps a lot, but since I didn't write it down, here are photos of mine that you can reverse engineer or use to confirm your wiring.  

![IMG_2423](https://github.com/user-attachments/assets/9b654426-3f7e-49e0-96d7-06588061178e)
![IMG_2422](https://github.com/user-attachments/assets/6becee1b-6d31-4799-8c1e-45f2314eb4d4)

**Reply by jimdinunzio** [reply]
I like that model of a PET! Where can I get one? I have a 3d printer if it helps, but the details must be from some other means.

**Comment by rbergen** [comment, marked as answer]
So I dug up the cable specs that I literally wrote two years ago **to the day**:

The pinout of the connector is as follows (edited to take out my personal projection of what's pin 1 to 4 on the grove connector):

- User port pin M (PA2) connected to G33 on the M5StickC Plus' grove connector
- User port pins B (/FLAG2), C (D0) connected together to G32 on the M5StickC Plus' grove connector
- User port pins A, N, 12 (all ground) connected together to the ground pin on the M5StickC Plus's grove connector
- The cable connected to the remaining pin on the M5StickC Plus grove connector is not used

And to prove the "two years to the day" isn't made up, here's a screen snippet of the e-mail I sent at the time:

![image](https://github.com/user-attachments/assets/8f37d7fe-7491-4362-9389-a0adb0f8a6c9)

**Reply by jimdinunzio** [reply]
Great! Thanks, Rutger! 

It is the same pin out that I suspected and posted earlier for the PET. 
For those interested, at least for A-N pins of the user port, pin names were changed between the PET and the C64, but the locations of the pins and use of them as GPIO lines seemed to remain the same.

C64 User Port Pin Out:
![image](https://github.com/user-attachments/assets/0c1b3c14-5136-46c3-9294-041e8a9844b9)

```
PET User Port Pin Out
+-----+-------------+-----+--------+
| PIN | SIGNAL      | PIN | SIGNAL |
+-----+-------------+-----+--------+
|  1  | GND         |  A  | GND    | --> N, GND from M5
|  2  | VIDEO       |  B  | CA1(Rx)| --> Tx (pin 3 of m5stick Grove conn., ESP pin 32)
|  3  | /SRQ IN     |  C  | PA0    | --> CA1
|  4  | /EOI        |  D  |  |1    |
|  5  | DIAG        |  E  |  |2    |
|  6  | #2CASSREAD  |  F  |  |3    |
|  7  | CASSWRITE   |  H  |  |4    |
|  8  | #1CASSREAD  |  J  |  |5    |
|  9  | VERT DRIVE  |  K  |  |6    |
| 10  | HORZ DRIVE  |  L  |  V7    |
| 11  | GRAPHIC     |  M  | CB2(Tx)| --> Rx (pin 4 of m5stick Grove conn., ESP pin 33)
| 12  | GND         |  N  | GND    | --> A (GND)
+----------------------------------+
```

**Reply by jimdinunzio** [reply]
I made two of them and both work fine.

**Reply by rbergen** [reply]
That's good to hear. Thanks for letting us know!

**Comment by davepl** [comment]
I can confirm that the same adapter works on the PET and C64, at least for me!  At least that’s what I remember - it’s a bit surprising since the CIA chips are different between the two, but I guess the pins were just remapped similarly enough that it “just works!"

**Reply by rbergen** [reply]
Yeah, as I recall from the time I was working on the C64 and PET serial drivers in the context of PETRock, the difference is that you drive the signals on the PET's VIA at a much lower level than the C64's CIA. To the point that you have to replace the most fundamental interrupt handler on the PET, which in turn makes that on that machine you have to implement your own keyboard handler to have a chance of making serial work at an acceptable baud rate.

Beyond all that, it's indeed "just" a matter of making the same signals from an RS232 perspective end up on the same pins on the User port.

**Comment by davepl** [comment]
’Twas a gift from a friend… I suspect ETSY, but don’t know!
